### PR TITLE
Add MatrixFreeOperators::MassOperator

### PR DESCRIPTION
**Pull request by masterleinad**

### CODE REVIEW

File: include/deal.II/matrix_free/operators.h [hunk 1]
```diff
@@ -288,6 +288,48 @@ namespace MatrixFreeOperators
 
 
 
+  /**
+   * This class implements the operation of the action of a mass matrix
+   * and additionally the action of a lumped mass matrix that can be used for
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: i would reword the second part. The class does not implement an action of lumped mass matrix, in that you can't do `vmult` with it. Instead, the diagonal is computed by lumping all elements, which is one of many possible approaches. I would just move this info to `compute_diagonal()` and not mention it here.

File: include/deal.II/matrix_free/operators.h [hunk 2]
```diff
@@ -288,6 +288,48 @@ namespace MatrixFreeOperators
...
+   * preconditioning. This class is used in VectorTools::project_distributed().
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: Probably not important where it is used.

File: include/deal.II/matrix_free/operators.h [hunk 3]
```diff
@@ -288,6 +288,48 @@ namespace MatrixFreeOperators
...
+
+    /**
+     * For preconditioning, we store a lumped mass matrix at the diagonal entries.
+     */
+    virtual void compute_diagonal ();
+
+  private:
+    /**
+     * Applies the mass matrix operation on an input array. It is
+     * assumed that the passed input and output arrays are of correct size,
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: you don't pass, arrays but LA::distr::Vectors and the size feels wrong, it looks like the description is out of place.

File: include/deal.II/matrix_free/operators.h [hunk 4]
```diff
@@ -288,6 +288,48 @@ namespace MatrixFreeOperators
...
+   */
+  template <int dim, int fe_degree, int n_components = 1, typename number = double>
+  class MassOperator : public Base<dim, number>
+  {
+  public:
+
+    /**
+     * Constructor. Initializes the MatrixFree object.
+     */
+    MassOperator (const MatrixFree<dim,number> &data);
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: Please, make it consistent with the base class, namely use default constructor. Otherwise we don't differentiate between initialization at fine scale (only MatrixFree data is neede) vs initialization at the GMG level (MatrixFree data + MGConstrainedDoFs + level). (see Base class).

File: include/deal.II/matrix_free/operators.h [hunk 5]
```diff
@@ -717,6 +759,92 @@ namespace MatrixFreeOperators
...
+
+  //-----------------------------MassOperator----------------------------------
+
+  template <int dim, int fe_degree, int n_components, typename number>
+  MassOperator<dim, fe_degree, n_components, number>::
+  MassOperator (const MatrixFree<dim,number> &data)
+    :
+    Base<dim, number>()
+  {
+    Base<dim, number>::initialize(data);
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: i suggest to remove this.

**Comment by davydden** [comment]
@kronbichler you probably want to look at this PR as well.

File: include/deal.II/lac/precondition.h [hunk 6]
```diff
@@ -1842,7 +1842,8 @@ PreconditionChebyshev<MatrixType,VectorType>::initialize
               ExcMessage ("Need to set at least two iterations to find eigenvalues."));
 
       // set a very strict tolerance to force at least two iterations
-      ReductionControl control (data.eig_cg_n_iterations, 1e-35, 1e-10);
```
Comment (masterleinad) [inline include/deal.II/lac/precondition.h]: Well, not really. I just noticed that the test results contain output from this solve operation which is not expected to converge.

File: tests/matrix_free/mass_operator_01.cc [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                                                  local_dof_indices,
+                                                  sparse_matrix);
+        }
+  }
+  sparse_matrix.compress(VectorOperation::add);
+
+  sparse_matrix.vmult (ref, in);
+  out -= ref;
+  const double diff_norm = out.linfty_norm();
+
```
Comment (davydden) [inline tests/matrix_free/mass_operator_01.cc]: i would also add a test of you muss lumping, you can use 
```
const_iterator TrilinosWrappers::SparseMatrix::begin(const size_type r) const
```
and `end(r)` to access elements in the row. Probably there is a better way to do that, but it should work.

File: include/deal.II/lac/precondition.h (hunk 6 above)
Comment (davydden) [inline include/deal.II/lac/precondition.h]: oh, i see.

File: tests/matrix_free/mass_operator_02.cc [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+// this tests the correctness of matrix_diagonal_inverse of
```
Comment (davydden) [inline tests/matrix_free/mass_operator_02.cc]: how do you know they are correct, you don't compare it with anything.

**Comment by davydden** [comment]
LGTM, unless @kronbichler has further comments, please merge yourself tomorrow afternoon.

File: include/deal.II/matrix_free/operators.h [hunk 9]
```diff
@@ -717,6 +757,80 @@ namespace MatrixFreeOperators
...
+    Base<dim, number>::initialize_dof_vector(ones);
+    Base<dim, number>::inverse_diagonal_entries = 1.;
+    ones = 1.;
+    ones.update_ghost_values();
+    apply_add(Base<dim, number>::inverse_diagonal_entries, ones);
+
+    const unsigned int local_size = Base<dim, number>::inverse_diagonal_entries.local_size();
+    for (unsigned int i=0; i<local_size; ++i)
+      {
+        Base<dim, number>::inverse_diagonal_entries.local_element(i)
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: dont u need to do 
```
 for (unsigned int i=0; i<constrained_dofs.size(); ++i)
    Base<dim, number>::inverse_diagonal_entries.local_element(constrained_dofs[i]) = 1
```

Comment (kronbichler) [inline include/deal.II/matrix_free/operators.h]: Either set them to 1 after this loop or catch the case where the entry is less than say `100.*std::numeric_limits<number>::min()` (the entries should only be positive or zero if constrained) and put 1.

File: tests/matrix_free/mass_operator_02.cc (hunk 8 above)
Comment (masterleinad) [inline tests/matrix_free/mass_operator_02.cc]: The output files for different number of processes are the same. This is what I wanted to test here.

File: include/deal.II/matrix_free/operators.h (hunk 9 above)
Comment (masterleinad) [inline include/deal.II/matrix_free/operators.h]: No, I set all elements to one beforehand and the constrained ones are not touched during `vmult_add`.

File: tests/matrix_free/mass_operator_01.cc (hunk 7 above)
Comment (kronbichler) [inline tests/matrix_free/mass_operator_01.cc]: I guess that method is fine. The alternative is to get the `trilinos_matrix()` and from that extract the appropriate `LocalRowView`, but that method is not portable.

File: include/deal.II/matrix_free/operators.h (hunk 9 above)
Comment (masterleinad) [inline include/deal.II/matrix_free/operators.h]: @kronbichler I go with your second suggestion.

**Comment by davydden** [comment]
@masterleinad : so it looks like this is ready apart from moving out changes to ReductionControl, right?

File: include/deal.II/lac/precondition.h (hunk 6 above)
Comment (kronbichler) [inline include/deal.II/lac/precondition.h]: I will fix this with #3294, so no need for you @masterleinad to address this. Thanks for pointing things out.

**Comment by masterleinad** [comment]
@davydden From my point of view this is now ready. I removed the part with ReductionControl and improved the second test.

**Comment by Rombur** [comment]
I quickly went through it and it looks really nice. My only "comment" is that you use `number` as a template parameter while most of the time we use `Number`. But I like this PR :+1:

**Comment by davydden** [comment]
@Rombur i created an issue to address this renaming https://github.com/dealii/dealii/issues/3303 as it is the case in other places in this header.

**Comment by davydden** [comment]
@masterleinad thanks!
